[SDK-364] Internal Limits

turtledreams opened the pull request:
- Created 2 internal functions:

1. `truncateSingleValue` takes a single string or number and truncates it for a given limit, for example:
     `truncateSingleValue('abcd', 2)` will return `'ab'`
     and also it will create a log informing the user that the value is over the limit and will be truncated.
2. `truncateObject` takes an object and returns a new object with truncated key value pairs, for example:
     `truncateObject({'ab':'1234'}, 1, 2)` will return `{'a':'12'}`
     and also it will create a log informing the user that the key/value is over the limit and will be truncated.

- Used these functions to enforce internal limits where it seemed fit, especially segmentations and event creation.
- Limit values are set at init time, except the internal one (`maxStackTraceThreadCount`)
- Erased/Renamed `max_logs` with `max_breadcrumb_count`
- Added their respective comments except for `maxStackTraceThreadCount`
- Used regex to limit stack trace line limits.
- Performed unit tests on regex and functions individually. 
- Changed `srcElement` with `target` as it was deprecated
- Added changelog entry (was not sure about labeling, maybe potentially/minor breaking change)

ArtursKadikis commented on CHANGELOG.md:
```diff
@@ -8,6 +8,13 @@
     - `report_feedback` needs 'star-rating' consent only
 - !! Major breaking change !! Enabling offline mode or changing device ID without merging will now clear the current consent. Consent has to be given again after performing this action.
 - ! Minor breaking change ! 'change_id' will now not accept invalid device ID values. It will now reject null, undefined, values that are not of the type string and empty string values.
+- Multiple values now have a default limit adjustable at initialization:
```
mark this as minor breaking change

ArtursKadikis commented on lib/countly.js:
```diff
@@ -3316,6 +3368,53 @@
+     */
+    function truncateSingleValue(str, limit) {
+        var newStr = str;
+        if (str.length > limit && (typeof str === 'string' || typeof str === 'number')) {
```
Type check would probably need to come first. I think it would throw an exception if you would call ".length" on a number or boolean.

ArtursKadikis commented on lib/countly.js:
```diff
@@ -3316,6 +3368,53 @@
+        for (var key in obj) {
+            var newKey = key;
+            var newValue = obj[key];
+            if (key.length > keyLimit) {
```
type of key need to be checked

ArtursKadikis commented on lib/countly.js:
```diff
@@ -3316,6 +3368,53 @@
+                newKey = key.substring(0, keyLimit);
+                this.log("Key: " + key + " is longer than accepted length. It will be truncated.");
+            }
+            if (obj[key].length > valueLimit && (typeof obj[key] === 'string' || typeof obj[key] === 'number')) {
```
would probably need similar early type checking as described for "truncateSingleValue"

ArtursKadikis commented on lib/countly.js:
```diff
@@ -2311,6 +2351,11 @@
                 else {
                     error = err + "";
                 }
+                var errorRegex = new RegExp("(?!$|\n)([^\n]{" + this.maxStackTraceLineLength + "}(?!\n))", "g");
```
No need for regex. Let's keep it simple with loops and function calls.

ArtursKadikis reviewed (changes requested):
Leave "max_logs" in. To get the resulting maxBreadcrumbCount value, you would check both breadcrumb values. If one of them is set, you would use that value. If both of them are set, you would use "maxBreadcrumbCount". If none of them are set, you would use the default for "maxBreadcrumbCount".

"max_logs" would need to be deprecated.

All of these limits would need to be described in the doc.

ArtursKadikis commented on lib/countly.js:
```diff
@@ -183,6 +183,7 @@
             trackTime = true,
             startTime = getTimestamp(),
             lsSupport = true,
+            maxStackTraceThreadCount = 30,
```
why is this set?

ArtursKadikis reviewed (changes requested):
This also requires a sample app/page where this is tested. That would test all values to make sure that all possible limits are enforced.

Current  breadcrumb login does not work like discussed. It should check if 'max_logs' has a value set. If it has, it should use that. If both are set, it should use "maxBreadcrumbCount".

From line 921 to line 1025 all "this.debug" calls should be replaced with "self.debug" as they would be linking to the context of surounding object where they are set.

To reduce the chance of potential future errors and consistency, use "self.XXX" instead of "this.XXX" when referencing "global" things. This currently looks to be only required for "this.debug" references.

Both "truncateSingleValue" and "truncateObject" need to be simplified. They also miss a few things discussed previously. 
In the single value call you would do the "number" check first, and if it would be a number, then you would only cast it to string and nothing else (basicly only lines 3560 - 3561. After that it would check the string type check. This way you get rid of duplicate code and make it more compact. 
In "truncateObject" for both the key and value you just call "truncateSingleValue", as that already has all of the required functionality.

ArtursKadikis reviewed (changes requested):
This also requires a sample app/page where this is tested. That would test all values to make sure that all possible limits are enforced.

Current  breadcrumb login does not work like discussed. It should check if 'max_logs' has a value set. If it has, it should use that. If both are set, it should use "maxBreadcrumbCount".

From line 921 to line 1025 all "this.debug" calls should be replaced with "self.debug" as they would be linking to the context of surounding object where they are set.

To reduce the chance of potential future errors and consistency, use "self.XXX" instead of "this.XXX" when referencing "global" things. This currently looks to be only required for "this.debug" references.

Both "truncateSingleValue" and "truncateObject" need to be simplified. They also miss a few things discussed previously. 
In the single value call you would do the "number" check first, and if it would be a number, then you would only cast it to string and nothing else (basicly only lines 3560 - 3561. After that it would check the string type check. This way you get rid of duplicate code and make it more compact. 
In "truncateObject" for both the key and value you just call "truncateSingleValue", as that already has all of the required functionality.

turtledreams commented:
- Added example file
- Fixed and tested breadcrumbs
- Usage of `this` is fixed
- `typeof` check is now shorter
- `truncateSingleValue` is now re-used in `truncateObject`
- All tested with example

ArtursKadikis reviewed (changes requested):
"maxBreadcrumbCount" value setting logic should be moved to init.

Test should be tweaked:
* add explicit event recording that will be truncated
* add user profile value recording that will be truncated
* for user profiless use also some property modiffier

Add square brackets to printed values

turtledreams commented:
- `maxBreadcrumbCount`  setting logic moved to init
- Added user profile value recording 
- Added all property modifiers
- Added square brackets to printed values
- Fixed an issue where only internal events were passing `add_event`
- Changelog entry set to 21.11.0

ArtursKadikis commented on lib/countly.js:
```diff
@@ -729,6 +744,7 @@
         function add_cly_events(event) {
             //ignore bots
             if (self.ignore_visitor) {
+                log('a');
```
seems like a dev log which should be removed

ArtursKadikis commented on lib/countly.js:
```diff
@@ -738,6 +754,8 @@
             }
 
             if (!event.count) {
+                log('b');
```
seems like a dev log which should be removed

ArtursKadikis commented on lib/countly.js:
```diff
@@ -689,31 +697,38 @@
-                break;
-            default:
-                respectiveConsent = false;
+            if (event.key in internalEventKeyEnums) {
```
should be reworked, as this would add a performance penalty

ArtursKadikis commented on examples/example_internal_limits.html:
```diff
@@ -0,0 +1,168 @@
+      function recordError() {
+        //create new error object with stack key
+        var error = {
+          stack: "Enter your\n error here\n separated\n with newlines\n",
```
add more lines so that it is more pronounced that it is truncated

ArtursKadikis commented on examples/example_internal_limits.html:
```diff
@@ -0,0 +1,168 @@
+      //add user details
+      function addUserDetails() {
+        Countly.user_details({
+          name: "Arturs Sosins",
```
probably better us some other name :D

ArtursKadikis commented on examples/example_internal_limits.html:
```diff
@@ -0,0 +1,168 @@
+      function addUserDetails() {
+        Countly.user_details({
+          name: "Arturs Sosins",
+          username: "ar2rsawseen",
```
same with this username

ArtursKadikis commented on examples/example_internal_limits.html:
```diff
@@ -0,0 +1,168 @@
+          phone: "+37112345678",
+          //Web URL pointing to user picture
+          picture:
+            "https://pbs.twimg.com/profile_images/1442562237/012_n_400x400.jpg",
```
and also this picture link :D

ArtursKadikis commented on examples/example_internal_limits.html:
```diff
@@ -0,0 +1,168 @@
+            "https://pbs.twimg.com/profile_images/1442562237/012_n_400x400.jpg",
+          gender: "M",
+          byear: 1987, //birth year
+          custom: {
```
have 10 custom keys

ArtursKadikis commented on examples/example_internal_limits.html:
```diff
@@ -0,0 +1,168 @@
+        });
+      }
+      //modify user data
+      function set() {
```
these modification calls should be reworked so that each of them breaks the set limits

ArtursKadikis commented on examples/example_internal_limits.html:
```diff
@@ -0,0 +1,168 @@
+        Countly.add_event({
+          key: "Enter your key here",
+          count: 1,
+          segmentation: {
```
this should have 10 segmentation values

ArtursKadikis commented:
@ar2rsawseen and @itsiprikshit , could you please have a look to see if you notice any issues?
